src/components/Select/hooks.js

```diff
@@ -0,0 +1,211 @@
+    );
+
+  const setOption = Array.isArray(value)
+    ? (option, knownIndex) => {
+        if (disabled) return;
+        const index =
+          knownIndex ||
+          findExact(_options, getOptionLabel, getOptionValue, option);
+        let newSelectedOptions = [];
+        if (
+          value.find(
+            (selectedOption) =>
+              getOptionLabel(selectedOption) === getOptionLabel(option)
+          )
+        )
+          newSelectedOptions = removeSelectedOptionFilter(option);
+        else if (!value.length) {
+          newSelectedOptions = [option];
+        } else {
+          newSelectedOptions = [...value, option];
+        }
+        setFocusedOptionIndex(index);
+        if (searchable && onInputChange && inputValue) onInputChange("");
+        onChange(newSelectedOptions);
+      }
+    : (option, knownIndex) => {
+        const index =
+          knownIndex ||
+          findExact(_options, getOptionLabel, getOptionValue, option);
+        resetMenu(index);
+        onChange(option);
+      };
+
+  const [containerRef] = useClickOutside(resetMenu);
+
+  const selectOption = () => {
+    if (!menuOpened) return;
+    const optionIndex = focusedOptionIndex;
+    const optionByIndex = _options[optionIndex];
+    if (_options[optionIndex].onClick !== undefined) {
+      _options[optionIndex].onClick();
+      return;
+    }
+    setOption(optionByIndex, optionIndex);
+  };
+
+  const openMenu = () =>
+    setMenuOpened((menuOpened) => !disabled && !menuOpened);
+
+  const cancelSelection = (e) => {
+    if (disabled) return;
+    onChange(blankValue);
+    resetMenu();
+    e.stopPropagation();
+  };
+
+  const onSearch = (e) => {
+    const searchValue = e.target.value;
+    if (onInputChange) onInputChange(searchValue);
+  };
+
+  const transitions = useTransition(menuOpened, null, {
+    from: { opacity: 0 },
+    enter: { opacity: 1 },
+    leave: { opacity: 0 },
+    duration: 100
+  });
+
+  return {
+    focusedOptionIndex,
+    setFocusedOptionIndex,
+    menuOpened,
+    setMenuOpened,
+    selectOption,
+    openMenu,
+    containerRef,
+    resetMenu,
+    cancelSelection,
+    onSearch,
+    transitions,
+    _options,
+    setOptions,
+    setOption
+  };
+}
+
+export function useHandleKeyboardHookBasicParameters(
+  menuOpened,
+  options,
+  focusedOptionIndex,
+  setFocusedOptionIndex,
+  inputValue,
+  onInputChange,
+  searchable = true
+) {
+  const onTypeArrowDownHandler = () => {
+    if (!menuOpened) return;
+    const maxOptionIndex = options.length - 1;
+    const newIndex =
+      focusedOptionIndex === maxOptionIndex ? 0 : focusedOptionIndex + 1;
+    setFocusedOptionIndex(newIndex);
+  };
+
+  const onTypeArrowUpHandler = () => {
+    if (!menuOpened) return;
+    const maxOptionIndex = options.length - 1;
+    const newIndex =
+      focusedOptionIndex === 0 ? maxOptionIndex : focusedOptionIndex - 1;
+    setFocusedOptionIndex(newIndex);
+  };
```

Comment:
the only difference between `onTypeArrowDownHandler` and `onTypeArrowUpHandler` is the newIndex condition. Can you pass a `isUpDirection` as an argument, and use a generic `onTypeArrowHandler(isUpDirection)` instead?